corehq/apps/cleanup/management/commands/fix_forms_with_missing_xmlns.py

```diff
@@ -0,0 +1,233 @@
+import uuid
+from itertools import chain
+
+from couchdbkit import ResourceNotFound
+
+from corehq.apps.app_manager.exceptions import FormNotFoundException
+from corehq.apps.app_manager.models import Application
+from corehq.apps.app_manager.util import get_correct_app_class
+from corehq.apps.app_manager.xform import XForm, parse_xml
+from corehq.apps.es import FormES
+from corehq.apps.es.filters import NOT, doc_type
+from corehq.apps.es.forms import xmlns
+from corehq.util.couch import IterDB
+from corehq.util.quickcache import quickcache
+from couchforms.const import ATTACHMENT_NAME
+from couchforms.models import XFormInstance
+from dimagi.utils.couch.database import iter_docs
+from django.core.management.base import BaseCommand
+from optparse import make_option
+
+
+ONE_HOUR = 60 * 60
+
+
+class Command(BaseCommand):
+    help = 'Fix forms with "undefined" xmlns'
+    args = '<log_path>'
+
+    option_list = BaseCommand.option_list + (
+        make_option(
+            '--dry-run',
+            action='store_true',
+            dest='dry_run',
+            default=False,
+            help="Don't do the actual modifications, but still log what would be affected"
+        ),
+    )
+
+    def handle(self, *args, **options):
+
+        dry_run = options.get("dry_run", True)
+        log_path = args[0].strip()
+
+        new_xmlnss = {}
+        num_fixed = 0
+        total, submissions = get_submissions_without_xmlns()
+
+        with open(log_path, "w") as f:
+            xform_db = IterDB(XFormInstance.get_db())
+            app_db = IterDB(Application.get_db())
+            with xform_db as xform_db:
+                with app_db as app_db:
+                    for i, xform_instance in enumerate(submissions):
+                        self._print_progress(i, total, num_fixed)
+                        if xform_instance.app_id in new_xmlnss:
+                            num_fixed += 1
+                            # We've already generated a new xmlns for this app
+                            set_xmlns_on_submission(
+                                xform_instance,
+                                new_xmlnss[xform_instance.app_id],
+                                xform_db,
+                                dry_run,
+                            )
+                            f.write("Set new xmlns on submission {}\n".format(xform_instance._id))
+                        else:
+                            app = Application.get(xform_instance.app_id)
+                            forms_without_xmlns = get_forms_without_xmlns(app)
+                            if len(forms_without_xmlns) == 1:
+                                form = forms_without_xmlns[0]
+                                if not xforms_with_real_xmlns_possibly_exist(app._id, form):
```

Comment:
Hmm. I don't think I thought about that one. There is a validation on saving the xform that forces the xmlns to be unique within the app, but if one was fixed in the time that the bug was live that could cause another form to have an `undefined` xmlns. Sounds like comparing form name would be a good idea.